devito/operator/operator.py

```diff
@@ -1268,6 +1321,89 @@ def nbytes_avail_mapper(self):
             nproc = 1
         mapper[host_layer] = int(ANYCPU.memavail() / nproc)
 
+        for layer in (host_layer, device_layer):
+            mapper[layer] -= self.nbytes_consumed_operator.get(layer, 0)
+
+        mapper = {k: int(v) for k, v in mapper.items()}
+
+        return mapper
+
+    @cached_property
+    def nbytes_consumed(self):
+        """Memory consumed by all objects in the operator"""
+        mem_locations = (
+            self.nbytes_consumed_function,
+            self.nbytes_consumed_array,
+            self.nbytes_consumed_memmapped
+        )
+        return {layer: sum(loc[layer] for loc in mem_locations) for layer in _layers}
+
+    @cached_property
+    def nbytes_consumed_operator(self):
+        """Memory consumed by objects allocated within the operator"""
+        mem_locations = (
+            self.nbytes_consumed_array,
+            self.nbytes_consumed_memmapped
+        )
+        return {layer: sum(loc[layer] for loc in mem_locations) for layer in _layers}
+
+    @cached_property
+    def nbytes_consumed_function(self):
+        """
+        Memory consumed on both device and host by Functions in the
+        corresponding operator.
+        """
+        def get_nbytes(obj):
+            if obj.is_regular:
+                nbytes = obj.nbytes
+            else:
+                nbytes = obj.nbytes_max
+
+            # Could nominally have symbolic nbytes at this point
+            if isinstance(nbytes, SympyBasic):
+                return subs_op_args(nbytes, self)
+            else:
+                return nbytes
+
+        host = 0
+        device = 0
+
+        # Symbols in the operator which may or may not carry data
```

Comment:
These `op_symbols` lines could be a private, cached property, since they are re-executed in another method below too